### PR TITLE
Add 'class' URI scheme for dynamic test sources

### PR DESCRIPTION
**Pull request by therealryan**
## Overview

 * Adding `ClassSource.from(URI)` method, which accepts URIs with a new `class` scheme. The URI details the fully qualified class name and optional `line` and `column` query parameters.
 * Hooking that new scheme into `TestFactoryTestDescriptor.fromUri(URI)`, so that such URIs can be supplied when creating dynamic tests
 *  Associated unit test and documentation updates

This enables dynamic tests instances to be located using the information available in a StackTraceElement, so it becomes possible to automatically provide test locations by walking the call stack.

These changes are largely a subset of #2431. To the extent that they work, credit goes to @ST-DDT. To the extent that they're broken, blame falls upon me.

Resolves issue: #2635

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Review by ST-DDT — APPROVED** [review]

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 1]
```diff
@@ -1830,6 +1830,10 @@ implementations.
   refer to the Javadoc for `DiscoverySelectors.selectMethod(String)` for the supported
   formats for a FQMN.
 
+`ClassSource` ::
+  If the `URI` contains the `class` scheme and the fully qualified class name --
+  for example, `class:/org.junit.Foo?line=42`.
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:1835]: ❓ Is there a technical reason why we require a leading slash here? I think `class:org.junit.Foo?line=42` would be more consistent with the `method` scheme.

Comment (therealryan) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:1835]: It's the distinction between opaque and hierarchical URIs - paths that do not start with a slash are opaque, and are more work to extract the query from, e.g.:
```
Stream.of(
    "scheme:path?query#fragment",
    "scheme:/path?query#fragment" )
    .forEach( s -> {
	    try {
		    URI uri = new URI( s );
		    System.out.println( String.format(
		        "uri '%s' scheme='%s' ssp='%s' path='%s' query='%s' fragment='%s'",
		        uri,
		        uri.getScheme(),
		        uri.getSchemeSpecificPart(),
		        uri.getPath(),
		        uri.getQuery(),
		        uri.getFragment() ) );
	    }
	    catch ( Exception e ) {
		    e.printStackTrace();
	    }
    } );
```
produces
```
uri 'scheme:path?query#fragment' scheme='scheme' ssp='path?query' path='null' query='null' fragment='fragment'
uri 'scheme:/path?query#fragment' scheme='scheme' ssp='/path?query' path='/path' query='query' fragment='fragment'
```

I do agree that it's worth the effort to get rid of that leading slash though, I'll make the change

File: platform-tests/src/test/java/org/junit/platform/engine/support/descriptor/ClassSourceTests.java [hunk 2]
```diff
@@ -89,6 +94,15 @@ void classSourceFromReferenceAndFilePosition() {
...
 
+	@Test
+	void classSourceFromURI() throws URISyntaxException {
+		var source = ClassSource.from(new URI("class:java.lang.Object?line=42"));
+
+		assertThat(source.getJavaClass()).isEqualTo(Object.class);
+		assertThat(source.getPosition()).isNotEmpty();
+		assertThat(source.getPosition().get().getLine()).isEqualTo(42);
+	}
+
```
Comment (marcphilipp) [inline platform-tests/src/test/java/org/junit/platform/engine/support/descriptor/ClassSourceTests.java:105]: Please also add a test for an URI without a query suffix.

**Review by marcphilipp — CHANGES_REQUESTED** [review]

**Review by marcphilipp — APPROVED** [review]
LGTM but Spotless will complain about the formatting. Please run `./gradlew spotlessApply` and commit the result.

**Comment by therealryan** [comment]
I couldn't get `./gradlew spotlessApply` to work for me, I got failures of the form:
```
> Task :junit-jupiter-api:spotlessJava FAILED
Step 'removeUnusedImports' found problem in 'junit-jupiter-api\src\main\java\org\junit\jupiter\api\AfterAll.java':
null
java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:567)
        at com.diffplug.spotless.java.GoogleJavaFormatStep$State.lambda$constructRemoveUnusedFunction$3(GoogleJavaFormatStep.java:190)
        at com.diffplug.spotless.java.GoogleJavaFormatStep$State.lambda$createRemoveUnusedImportsOnly$1(GoogleJavaFormatStep.java:167)
...
a very, very long stacktrace
...
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
        at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.lang.IllegalAccessError: class com.google.googlejavaformat.java.RemoveUnusedImports (in unnamed module @0x7cbdd845) cannot access class com.sun.tools.javac.util.Context (in module jdk.compiler) because module jdk.compiler does not export com.sun.tools.javac.util to unnamed module @0x7cbdd845
        at com.google.googlejavaformat.java.RemoveUnusedImports.removeUnusedImports(RemoveUnusedImports.java:188)
        ... 149 more

FAILURE: Build completed with 3 failures.
```

That said, I applied the eclipse formatter settings manually to the code changes, fixed an unencoded ampersand that javadoc was complaining about in those builds, and slightly clarified the wording while I was there.

**Comment by marcphilipp** [comment]
Sounds like you were running with JDK 16, correct? I recently fixed that in faa56b1cf7edac42b164dacba924cdcf47b6ea00.

**Comment by marcphilipp** [comment]
@therealryan Thanks a lot for your contribution! 👍

**Pull request merged**